ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -0,0 +1,560 @@
+  // insertion
+  std::vector<ddog_prof_ValueType> samplers;
+  auto get_value_idx = [&samplers](std::string_view value,
+                                   std::string_view unit) {
+    size_t idx = samplers.size();
+    samplers.push_back({to_slice(value), to_slice(unit)});
+    return idx;
+  };
+
+  // Check which samplers were enabled by the user
+  if (type_mask & ProfileType::CPU) {
+    val_idx.cpu_time = get_value_idx("cpu-time", "nanoseconds");
+    val_idx.cpu_count = get_value_idx("cpu-samples", "count");
+  }
+  if (type_mask & ProfileType::Wall) {
+    val_idx.wall_time = get_value_idx("wall-time", "nanoseconds");
+    val_idx.wall_count = get_value_idx("wall-samples", "count");
+  }
+  if (type_mask & ProfileType::Exception) {
+    val_idx.exception_count = get_value_idx("exception-samples", "count");
+  }
+  if (type_mask & ProfileType::LockAcquire) {
+    val_idx.lock_acquire_time =
+        get_value_idx("lock-acquire-wait", "nanoseconds");
+    val_idx.lock_acquire_count = get_value_idx("lock-acquire", "count");
+  }
+  if (type_mask & ProfileType::LockRelease) {
+    val_idx.lock_release_time =
+        get_value_idx("lock-release-hold", "nanoseconds");
+    val_idx.lock_release_count = get_value_idx("lock-release", "count");
+  }
+  if (type_mask & ProfileType::Allocation) {
+    val_idx.alloc_space = get_value_idx("alloc-space", "bytes");
+    val_idx.alloc_count = get_value_idx("alloc-samples", "count");
+  }
+  if (type_mask & ProfileType::Heap) {
+    val_idx.heap_space = get_value_idx("heap-space", "bytes");
+  }
+
+  values.resize(samplers.size());
+  std::fill(values.begin(), values.end(), 0);
+
+  ddog_prof_Period default_sampler = {
+      samplers[0], 1}; // Mandated by pprof, but probably unused
+  ddog_profile = ddog_prof_Profile_new({&samplers[0], samplers.size()},
+                                       &default_sampler, nullptr);
+
+  // Prepare for use
+  reset();
+
+  // Initialize the size for buffers
+  locations.reserve(2040);
+  lines.reserve(2048);
+  strings.reserve(8192);
+}
+
+Profile::~Profile() { ddog_prof_Profile_drop(ddog_profile); }
+
+bool Profile::reset() {
+  if (!ddog_prof_Profile_reset(ddog_profile, nullptr)) {
+    errmsg = "Unable to reset profile";
+    return false;
+  }
+  return true;
+}
+
+bool Profile::start_sample(unsigned int nframes) {
+  strings.clear();
+  clear_buffers();
+  this->nframes = nframes;
+  return true;
+}
+
+void Profile::push_frame_impl(std::string_view name, std::string_view filename,
+                              uint64_t address, int64_t line) {
+
+  // Ensure strings are stored.
+  // Slightly wasteful since it requires allocating another string.
+  auto insert_or_get = [&](std::string_view sv) -> std::string_view {
+    std::string str(sv);
+    auto [it, _] = strings.insert(std::move(str));
+    return *it;
+  };
+  name = insert_or_get(name);
+  filename = insert_or_get(filename);
+
+  lines.push_back({
```

Comment:
Bound checks are important, though they will come with the change in storage container I assume.

Reply:
This was changed in 61f10cc50 and subsequently had to be made more strict.